### PR TITLE
dialects: (emitc) Add EmitC dialect with the EmitC_ArrayType

### PR DESCRIPTION
**Pull request by Jimmy2027**
Register the emitc dialect in xdsl/dialects/__init__.py and add the EmitC Array Type as a first step towards adding the full [emitc dialect](https://mlir.llvm.org/docs/Dialects/EmitC/)

### CODE REVIEW

**Review by superlopuh — COMMENTED** [review]
Can you please add a test to check a roundtrip with mlir, like `tests/filecheck/mlir-conversion/with-mlir/dialects/arith/arith_attrs.mlir`, Also can you please make your test similar to that one (but omitting the implied builtin.module)

File: xdsl/dialects/emitc.py
Comment (superlopuh) [inline xdsl/dialects/emitc.py:1]: Can you please add a docstring, with a link to the linalg implementation, and a description? You can see what the generated docs site looks like by clicking on the readthedocs CI job.

Comment (superlopuh) [reply xdsl/dialects/emitc.py:1]: Sorry, I mistyped, I meant the MLIR dialect definition.

Comment (Jimmy2027) [reply xdsl/dialects/emitc.py:1]: ah okay makes more sense :) will add that!

File: xdsl/dialects/emitc.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        return self.element_type
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> list[Attribute]:
+        parser.parse_punctuation("<", " in array attribute")
+        shape, type = parser.parse_ranked_shape()
+        shape = ArrayAttr([IntAttr(dim) for dim in shape])
+        parser.parse_punctuation(">", " in array attribute")
+
+        return [shape, type]
```
Comment (superlopuh) [inline xdsl/dialects/emitc.py:70]:
```suggestion
    def parse_parameters(cls, parser: AttrParser) -> list[Attribute]:
        with parser.in_angle_brackets():
            shape, type = parser.parse_ranked_shape()
            return ArrayAttr(IntAttr(dim) for dim in shape), type
```

File: xdsl/dialects/emitc.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print(
+            "<",
+            "x".join([*[str(e.data) for e in self.shape], *[str(self.element_type)]]),
+        )
+        printer.print(">")
```
Comment (superlopuh) [inline xdsl/dialects/emitc.py:77]:
```suggestion
        with printer.in_angle_brackets():
            printer.print_list(
                self.shape,
                lambda dim: printer.print_string(f"{dim.data}"),
                "x"
            )
            if self.shape:
                printer.print_string("x")
            printer.print_attribute(self.element_type)
```

File: tests/filecheck/dialects/emitc/emitc_types.mlir [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+// Test the emitc array type
+// CHECK: func.func public @identity_array(%input : !emitc.array<2x2xf32>) -> !emitc.array<2x2xf32> {
+// CHECK-NEXT:   func.return %input : !emitc.array<2x2xf32>
+// CHECK-NEXT: }
+func.func public @identity_array(
+  %input: !emitc.array<2x2xf32>
+) -> !emitc.array<2x2xf32> {
+  
+  func.return %input : !emitc.array<2x2xf32>
+}
```
Comment (superlopuh) [inline tests/filecheck/dialects/emitc/emitc_types.mlir:12]: Why not just a "test.op" with an attribute in a dictionary?

Comment (superlopuh) [reply tests/filecheck/dialects/emitc/emitc_types.mlir:12]: Also, can you please add an MLIR roundtrip test, also with a test op

Comment (Jimmy2027) [reply tests/filecheck/dialects/emitc/emitc_types.mlir]: sorry messed up my update this morning, should be there now

File: tests/filecheck/dialects/emitc/emitc_types.mlir [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+"test.op"() {"array_attr" = !emitc.array<10xi32>}: ()->()
+// CHECK: array_attr = !emitc.array<10xi32>
+
+// Test EmitC_ArrayType with a different 1D shape and f64 element type
+"test.op"() {"array_attr" = !emitc.array<5xf64>}: ()->()
+// CHECK: array_attr = !emitc.array<5xf64>
+
+// Test EmitC_ArrayType with a 3D shape and i1 element type
+"test.op"() {"array_attr" = !emitc.array<3x4x5xi1>}: ()->()
+// CHECK: array_attr = !emitc.array<3x4x5xi1>
```
Comment (superlopuh) [inline tests/filecheck/dialects/emitc/emitc_types.mlir:17]: This is fine but I'd recommend something a little less verbose since we'll be adding more things:
```suggestion

// CHECK: f32_2D = !emitc.array<4x2xf32>
// CHECK-SAME: i32_1D = !emitc.array<10xi32>
// CHECK-SAME: f64_1D = !emitc.array<5xf64>
// CHECK-SAME: i1_3D = !emitc.array<3x4x5xi1>
"test.op"() {
  f32_2D = !emitc.array<4x2xf32>,
  i32_1D = !emitc.array<10xi32>,
  f64_1D = !emitc.array<5xf64>,
  i1_3D = !emitc.array<3x4x5xi1>,
}: ()->()
```
Also, MLIR puts the checks before, which I've tried to do in new test files.

Comment (Jimmy2027) [reply tests/filecheck/dialects/emitc/emitc_types.mlir]: right, thanks!

**Comment by Jimmy2027** [comment]
trailing comma was the issue :)

File: xdsl/dialects/emitc.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        with parser.in_angle_brackets():
+            shape, type = parser.parse_ranked_shape()
+            return ArrayAttr(IntAttr(dim) for dim in shape), type
+
+    def print_parameters(self, printer: Printer) -> None:
+        with printer.in_angle_brackets():
+            printer.print_list(
+                self.shape, lambda dim: printer.print_string(f"{dim.data}"), "x"
+            )
+            if self.shape:
```
Comment (superlopuh) [inline xdsl/dialects/emitc.py:73]: this is more of a nit, but it would be nice to have a test for an empty shape

**Review by superlopuh — APPROVED** [review]
Thank you!

File: xdsl/dialects/emitc.py (hunk 5 above)
Comment (Jimmy2027) [reply xdsl/dialects/emitc.py:73]: looks like this is a good point, an empty shape is not supported. I'm not sure if I can check for an error with lit, would it be better to add a pytest for this instead?

File: tests/filecheck/dialects/emitc/emitc_types.mlir [hunk 6]
```diff
@@ -0,0 +1,14 @@
+// RUN: XDSL_ROUNDTRIP
+
+// CHECK: f32_2D = !emitc.array<4x2xf32>
+// CHECK-SAME: i32_1D = !emitc.array<10xi32>
+// CHECK-SAME: f64_1D = !emitc.array<5xf64>
+// CHECK-SAME: i1_3D = !emitc.array<3x4x5xi1>
+// CHECK-SAME: EmitC_ArrayType shape must not be empty
```
Comment (superlopuh) [inline tests/filecheck/dialects/emitc/emitc_types.mlir:7]: I would recommend adding a separate test for this, something like `tests/filecheck/dialects/emitc/emitc_types_invalid.mlir`, and passing `--verify-diagnostics`, you can do a search in the project for similar tests.

File: xdsl/dialects/emitc.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    def __init__(
+        self,
+        shape: Iterable[int | IntAttr],
+        element_type: AttributeCovT,
+    ):
+        shape = ArrayAttr(
+            [IntAttr(dim) if isinstance(dim, int) else dim for dim in shape]
+        )
+        if len(shape.data) == 0:
+            raise ValueError("EmitC_ArrayType shape must not be empty")
```
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/dialects/emitc.py:52]: Actually I'm curious, what does `mlir-opt` raise as an error in this case?

[the remark this answers is not in the document]
Comment (Jimmy2027) [reply xdsl/dialects/emitc.py:52]: this :)
https://github.com/xdslproject/xdsl/actions/runs/15052998304/job/42312103517#step:9:279

[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/dialects/emitc.py:52]: Great, I would say let's use the mlir attribute name instead of the Python class name then we're GTG

File: xdsl/dialects/emitc.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    def get_element_type(self) -> AttributeCovT:
+        return self.element_type
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser):
+        with parser.in_angle_brackets():
+            shape, type = parser.parse_ranked_shape()
+            if not shape:
+                raise parser.raise_error("EmitC_ArrayType shape must not be empty")
```
Comment (superlopuh) [inline xdsl/dialects/emitc.py:69]: Oh we should have a separate parsing test also from verification, that is done with the `--parser-diagnostics` flag to `xdsl-op`, and we should add a `pytest` test for the verify exception

File: tests/filecheck/dialects/emitc/emitc_types.mlir (hunk 6 above)
Comment (superlopuh) [reply tests/filecheck/dialects/emitc/emitc_types.mlir:7]: I think we need to yoink this line now and then we'll be gtg

File: xdsl/dialects/emitc.py (hunk 8 above)
Comment (Jimmy2027) [reply xdsl/dialects/emitc.py]: the empty shape verifying is only tested in a pytest now, don't think it is possible from a lit test since that fails already at parsing

**Comment by superlopuh** [comment]
Thank you!
